test/test_tensordict.py

```diff
@@ -2829,13 +2829,57 @@ def test_update_nested_dict():
     t.update({"a": {"d": [[[1]] * 3] * 2}})
 
 
-@pytest.mark.parametrize("separator", [".", "-"])
-def test_unflatten_keys_collision(separator):
-    td = TensorDict(
-        {"a": [1, 2], f"c{separator}a": [1, 2], "c": TensorDict({"b": [1, 2]}, [])}, []
+@pytest.mark.parametrize("inplace", [True, False])
+@pytest.mark.parametrize("separator", [",", "-"])
+def test_flatten_unflatten_key_collision(inplace, separator):
+    td1 = TensorDict(
+        {
+            f"a{separator}b{separator}c": torch.zeros(3),
+            "a": {"b": {"c": torch.zeros(3)}},
+        },
+        [],
+    )
+    td2 = TensorDict(
+        {
+            f"a{separator}b": torch.zeros(3),
+            "a": {"b": torch.zeros(3)},
+            "g": {"d": torch.randn(3)},
+        },
+        [],
+    )
+    td3 = TensorDict(
+        {
+            f"a{separator}b{separator}c": torch.zeros(3),
+            "a": {"b": {"c": torch.zeros(3), "d": torch.zeros(3)}},
+        },
+        [],
+    )
+
+    td4 = TensorDict(
+        {
+            f"a{separator}b{separator}c{separator}d": torch.zeros(3),
+            "a": {"b": {"c": torch.zeros(3)}},
+        },
+        [],
+    )
+
+    td5 = TensorDict(
+        {f"a{separator}b": torch.zeros(3), "a": {"b": {"c": torch.zeros(3)}}}, []
     )
-    ref = TensorDict({"a": [1, 2], "c": TensorDict({"a": [1, 2], "b": [1, 2]}, [])}, [])
-    assert assert_allclose_td(td.unflatten_keys(separator), ref)
+
+    with pytest.raises(Exception):
+        _ = td1.flatten_keys(separator)
+        _ = td2.flatten_keys(separator)
+        _ = td3.flatten_keys(separator)
+
+        _ = td1.unflatten_keys(separator)
+        _ = td2.unflatten_keys(separator)
+        _ = td3.unflatten_keys(separator)
+        _ = td4.unflatten_keys(separator)
+        _ = td5.unflatten_keys(separator)
+
+    _ = td4.flatten_keys(separator)
+    _ = td5.flatten_keys(separator)
```

Comment:
This should raise an error no? `{"a": {"b" tensor}}` should collide with `{"a": {"b": {"c": tensor}}}`

Reply:
If my understanding is correct, as long as they flatten to different keys there's no collision, e.g. `td4` and `td5`  flatten to:

```
td4 = {
    "a.b.c.d": torch.zeros(3),
    "a.b.c": torch.zeros(3)
}

td5 = {
    "a.b": torch.zeros(3),
    "a.b.c": torch.zeros(3)
}
```

However, unflattening these cause a collision because the `c` key in `td4` has multiple values (`d` nested key and `torch.zeros(3)`. Same for the `b` key in `td5`.

Reply:
Right! Flatten works, unflatten shouldn't!

Reply:
Ok, then I think having the checks for flattening of `td4` and `td5` outside of a `pytest.raises` decorator is correct as is (we don't want these to raise errors).